core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -164,11 +167,25 @@ long memorySize() {
 
         updateIndexAndNormalize(engaged, newChunks);
 
-        for (Chunk chunk : engaged
-        ) {
+        for (Chunk chunk : engaged) {
+            ConcurrentSkipListSet<Iter> viewingIterators = chunk.getSignedIterators();
             chunk.release();
```

Comment:
shouldn't we wait with the release till after we set last key in iterators?

Reply:
No, the iterator does : 1) signin chunk. 2) check if chunk released
If we put the release after set lastkey there could be the following error:
1) iterator moves to new chunk. (not yet signin chunk)
2) rebalancer reads viewingIterator list but its still empty
3) iterator continues signin (will be lost) and checks if chunk marked release returns false
4) rebalancer marks chunk released and deletes chunk and buffers
5) iterator tries to read released buffer and all hell breaks loose.

If we first release then either the iterator will see the chunk is released, or the rebalancer will see the iterator in the list

Reply:
then this means the rebalancer code should
1) chunk release (currently line 172)
2) collect viewing iterators (currently line 171)
always reverse the order of the iterator at critical lines.

Reply:
The critical operations are 1) set/check lastKey 2) set/check release.
So leaving the release after collecting viewing iterators should be ok.

I think we should do the release first just to make the code clearer but i dont think its wrong this way.